### DISCUSSION TITLE
How to upgrade any RAG/search system from cosine to MizanSimilarity (and why you should)

### DISCUSSION THREAD

**Discussion by ahsanshaokat** [milvus-io/milvus#45796] [Ideas & Feature requests]
How to upgrade any RAG/search system from cosine to MizanSimilarity (and why you should)

Cosine similarity was designed for 2013-era NLP.
But today’s embeddings (LLMs like e5, MiniLM, mpnet, multilingual models, OCR data) produce wildly different magnitudes and information densities.

Cosine ignores all of this — leading to ranking failures.

MizanSimilarity and MizanDistance finally solve this by measuring balance, scale, and proportional fairness.

Here’s how to use them effectively.

1️⃣ Understanding the Metrics

✔ MizanSimilarity

Range: 0 → 1
Proportion-aware
Perfect for ranking
✔ MizanDistance

Range: 0 → ∞
Works with vector DBs & nearest-neighbor algorithms
They are mathematically inverse to each other — making them easy to optimize.

2️⃣ Concrete Code Example

from mizan_vector import mizan_similarity, mizan_distance 
import torch 
 
v1 = torch.randn(384) 
v2 = torch.randn(384) 
 
print("Similarity:", mizan_similarity(v1, v2)) 
print("Distance:",   mizan_distance(v1, v2)) 
3️⃣ RAG Re-Ranking (the #1 improvement)

Run first pass with cosine, then re-rank with Mizan:

reranked = sorted( 
    retrieved_chunks, 
    key=lambda x: mizan_similarity(query_emb, x["embedding"]), 
    reverse=True 
) 
This solves:

short text > long text errors
noisy OCR > clean text errors
embeddings from different domains collapsing
LLMs hallucinate less because retrieval becomes semantic, not geometric.

4️⃣ Using MizanMemoryStore

from mizan_vector import MizanMemoryStore 
 
store = MizanMemoryStore(metric="mizan") 
store.add("doc1", "...", embedding) 
results = store.search("meaning of balance") 
This gives you a full vector DB with Mizan scoring built-in.

5️⃣ Noise Robustness Advantage

v_clean = torch.randn(384) 
v_noisy = v_clean + torch.randn(384)*0.3 
 
print("Cosine:", cosine_similarity(v_clean, v_noisy)) 
print("Mizan:", mizan_similarity(v_clean, v_noisy)) 
Typical result:

cosine → 0.70–0.80
Mizan → 0.90+
Mizan sees noise as proportionally small, which is mathematically correct.

✔ When NOT to Use Mizan

dot-product-optimized models
embeddings already normalized to unit length
situations where magnitude must be ignored

Otherwise, Mizan is the superior default.
